internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -990,3 +996,13 @@ func FindStatusCondition(conditions []hivev1.ClusterInstallCondition, conditionT
 
 	return nil
 }
+
+// ensureOwnerRef sets the owner reference of ClusterDeployment on AgentClusterInstall
+func (r *ClusterDeploymentsReconciler) ensureOwnerRef(ctx context.Context, cd *hivev1.ClusterDeployment, ci *hiveext.AgentClusterInstall) error {
+
+	if err := controllerutil.SetOwnerReference(cd, ci, r.Scheme); err != nil {
+		r.Log.WithError(err).Error("error setting owner reference Agent Cluster Install")
+		return err
+	}
+	return r.Update(ctx, ci)
```

Comment:
This looks like it might always be doing a write, which will always re-reconcile, which will do another write, etc. If I'm not missing something you'd want to make see if something changed before calling Update.

Reply:
I checked and I don't see an infinite reconcile loop:

```
time="2021-05-12T13:03:07Z" level=info msg="Reconcile has been called for ClusterDeployment name=test-infra-cluster-assisted-installer namespace=assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:90"
time="2021-05-12T13:03:07Z" level=info msg="AgentClusterInstall does not exist for ClusterDeployment test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:122" AgentClusterInstall=test-agent-cluster-install
time="2021-05-12T13:03:24Z" level=info msg="Reconcile has been called for ClusterDeployment name=test-infra-cluster-assisted-installer namespace=assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:90"
time="2021-05-12T13:03:24Z" level=info msg="Creating a new clusterDeployment test-infra-cluster-assisted-installer assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).createNewCluster" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:578"
time="2021-05-12T13:03:24Z" level=info msg="Stored OCP version: 4.8.0" func="github.com/openshift/assisted-service/internal/versions.(*handler).AddOpenshiftVersion" file="/go/src/github.com/openshift/origin/internal/versions/versions.go:236" pkg=versions
time="2021-05-12T13:03:24Z" level=info msg="Register cluster: test-cluster with id a3898c66-c42c-4b4f-a783-1283bf4c91b8" func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).RegisterClusterInternal" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:322" cluster_id=a3898c66-c42c-4b4f-a783-1283bf4c91b8 go-id=517 pkg=Inventory request_id=
time="2021-05-12T13:03:24Z" level=info msg="HA mode is None, setting UserManagedNetworking to true and VipDhcpAllocation to false" func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).RegisterClusterInternal" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:365" cluster_id=a3898c66-c42c-4b4f-a783-1283bf4c91b8 go-id=517 pkg=Inventory request_id=
time="2021-05-12T13:03:24Z" level=info msg="Successfully registered cluster test-cluster with id a3898c66-c42c-4b4f-a783-1283bf4c91b8" func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).RegisterClusterInternal.func1" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:329" cluster_id=a3898c66-c42c-4b4f-a783-1283bf4c91b8 go-id=517 pkg=Inventory request_id=
time="2021-05-12T13:03:24Z" level=info msg="Reconcile has been called for ClusterDeployment name=test-infra-cluster-assisted-installer namespace=assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:90"
time="2021-05-12T13:03:24Z" level=info msg="update cluster a3898c66-c42c-4b4f-a783-1283bf4c91b8 with params: &{AdditionalNtpSource:<nil> APIVip:<nil> APIVipDNSName:<nil> BaseDNSDomain:<nil> ClusterNetworkCidr:<nil> ClusterNetworkHostPrefix:<nil> DisksSelectedConfig:[] HostsMachineConfigPoolNames:[] HostsNames:[] HostsRoles:[] HTTPProxy:<nil> HTTPSProxy:<nil> Hyperthreading:<nil> IngressVip:<nil> MachineNetworkCidr:0xc0005e24f0 Name:<nil> NoProxy:<nil> OlmOperators:[] PullSecret:<nil> ServiceNetworkCidr:<nil> SSHPublicKey:<nil> UserManagedNetworking:<nil> VipDhcpAllocation:<nil>}" func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).UpdateClusterInternal" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:1634" go-id=517 pkg=Inventory request_id=
time="2021-05-12T13:03:24Z" level=info msg="no infraEnv for the clusterDeployment test-infra-cluster-assisted-installer"
time="2021-05-12T13:03:24Z" level=info msg="Updated clusterDeployment assisted-installer/test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).updateIfNeeded" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:456"
```

Reply:
@dgoodwin What should I look for? We print on every Reconcile call.

Reply:
You'd see constant reconciles printed and high cpu use. I'm pretty confused how that's not happening, looks like a clear Update on every reconcile.

Reply:
We actually see the same behaviour with updating the conditions. We had an issue when we did update the timestamps every time, and now we update with the same conditions values but not getting a reconcile called again.

Reply:
I agree that's surprising. It may be useful to add a debug statement when the owner reference gets set to see if that's actually getting called and how much. Same in the ACI's map handler to make sure it's working as expected.

Reply:
Owner reference is set
```
  ownerReferences:
  - apiVersion: hive.openshift.io/v1
    kind: ClusterDeployment
    name: test-infra-cluster-assisted-installer
    uid: a7dc809b-3df2-46c5-b49e-f683f22bf8c9
```

Reply:
First create CD and then ACI:
No more Reconcile after these logs:
```
time="2021-05-13T07:22:58Z" level=info msg="Reconcile has been called for ClusterDeployment name=test-infra-cluster-assisted-installer namespace=assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:90"
time="2021-05-13T07:22:58Z" level=info msg="AgentClusterInstall does not exist for ClusterDeployment test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:122" AgentClusterInstall=test-agent-cluster-install
time="2021-05-13T07:23:18Z" level=info msg="Map ACI : assisted-installer test-agent-cluster-install CD ref name test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).SetupWithManager.func2" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:750"
time="2021-05-13T07:23:18Z" level=info msg="Reconcile has been called for ClusterDeployment name=test-infra-cluster-assisted-installer namespace=assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:90"
time="2021-05-13T07:23:18Z" level=info msg="Updating ACI owner ref: assisted-installer test-agent-cluster-install" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).ensureOwnerRef" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:1008"
time="2021-05-13T07:23:18Z" level=info msg="Map ACI : assisted-installer test-agent-cluster-install CD ref name test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).SetupWithManager.func2" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:750"
time="2021-05-13T07:23:18Z" level=info msg="Map ACI : assisted-installer test-agent-cluster-install CD ref name test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).SetupWithManager.func2" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:750"
time="2021-05-13T07:23:18Z" level=info msg="Creating a new clusterDeployment test-infra-cluster-assisted-installer assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).createNewCluster" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:578"
time="2021-05-13T07:23:18Z" level=info msg="Stored OCP version: 4.8.0" func="github.com/openshift/assisted-service/internal/versions.(*handler).AddOpenshiftVersion" file="/go/src/github.com/openshift/origin/internal/versions/versions.go:253" pkg=versions
time="2021-05-13T07:23:18Z" level=info msg="Register cluster: test-cluster with id cc6d2e09-9a61-48fc-8343-5a9bc04f7e20" func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).RegisterClusterInternal" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:323" cluster_id=cc6d2e09-9a61-48fc-8343-5a9bc04f7e20 go-id=945 pkg=Inventory request_id=
time="2021-05-13T07:23:18Z" level=info msg="HA mode is None, setting UserManagedNetworking to true and VipDhcpAllocation to false" func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).RegisterClusterInternal" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:366" cluster_id=cc6d2e09-9a61-48fc-8343-5a9bc04f7e20 go-id=945 pkg=Inventory request_id=
time="2021-05-13T07:23:19Z" level=info msg="Successfully registered cluster test-cluster with id cc6d2e09-9a61-48fc-8343-5a9bc04f7e20" func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).RegisterClusterInternal.func1" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:330" cluster_id=cc6d2e09-9a61-48fc-8343-5a9bc04f7e20 go-id=945 pkg=Inventory request_id=
time="2021-05-13T07:23:19Z" level=info msg="Map ACI : assisted-installer test-agent-cluster-install CD ref name test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).SetupWithManager.func2" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:750"
time="2021-05-13T07:23:19Z" level=info msg="Map ACI : assisted-installer test-agent-cluster-install CD ref name test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).SetupWithManager.func2" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:750"
time="2021-05-13T07:23:19Z" level=info msg="Reconcile has been called for ClusterDeployment name=test-infra-cluster-assisted-installer namespace=assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:90"
time="2021-05-13T07:23:19Z" level=info msg="Updating ACI owner ref: assisted-installer test-agent-cluster-install" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).ensureOwnerRef" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:1008"
time="2021-05-13T07:23:19Z" level=info msg="update cluster cc6d2e09-9a61-48fc-8343-5a9bc04f7e20 with params: &{AdditionalNtpSource:<nil> APIVip:<nil> APIVipDNSName:<nil> BaseDNSDomain:<nil> ClusterNetworkCidr:<nil> ClusterNetworkHostPrefix:<nil> DisksSelectedConfig:[] HostsMachineConfigPoolNames:[] HostsNames:[] HostsRoles:[] HTTPProxy:<nil> HTTPSProxy:<nil> Hyperthreading:<nil> IngressVip:<nil> MachineNetworkCidr:0xc000d3cf40 Name:<nil> NoProxy:<nil> OlmOperators:[] PullSecret:<nil> ServiceNetworkCidr:<nil> SSHPublicKey:<nil> UserManagedNetworking:<nil> VipDhcpAllocation:<nil>}" func="github.com/openshift/assisted-service/internal/bminventory.(*bareMetalInventory).UpdateClusterInternal" file="/go/src/github.com/openshift/origin/internal/bminventory/inventory.go:1635" go-id=945 pkg=Inventory request_id=
time="2021-05-13T07:23:19Z" level=info msg="no infraEnv for the clusterDeployment test-infra-cluster-assisted-installer"
time="2021-05-13T07:23:19Z" level=info msg="Updated clusterDeployment assisted-installer/test-infra-cluster-assisted-installer" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).updateIfNeeded" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:456"

```